parsec/mca/device/cuda/device_cuda_module.c

```diff
@@ -1052,19 +1046,31 @@ parsec_gpu_data_reserve_device_space( parsec_device_cuda_module_t* cuda_device,
                         /* If we are the owner of this tile we need to make sure it remains available for
                          * other tasks or we run in deadlock situations.
                          */
-			assert( temp_loc[j] == lru_gpu_elem );  /* dont understand how this cannot be true */
                         parsec_atomic_unlock( &oldmaster->lock );
                         goto find_another_data;
                     }
                 }
-                if( lru_gpu_elem->readers != 0 ) {
-                    /* Damn, another thread started to use this data (as source for an NVLINK transfer). */
+                /* There is still one last thing to ensure: if another accelerator uses this copy as a source
+                 * for a d2d transfer it will mark it by atomically increasing the readers. So, we need to
+                 * avoid altering the copy while they are using it, by protecting the access to the readers
+                 * with a cas.
+                 */
+                if( !parsec_atomic_cas_int32(&lru_gpu_elem->readers, 0, INT_MIN) ) {
+                    /* we can't use this copy, push it back */
+                    parsec_list_push_back(&gpu_device->gpu_mem_lru, &lru_gpu_elem->super);
+                    gpu_mem_lru_cycling = (NULL == gpu_mem_lru_cycling) ? lru_gpu_elem : gpu_mem_lru_cycling;  /* update the cycle detector */
+                    PARSEC_DEBUG_VERBOSE(20, parsec_gpu_output_stream,
+                                         "GPU[%s]:%s: Push back LRU-retrieved CUDA copy %p [readers %d, ref_count %d] original %p : Concurrent accesses",
+                                         gpu_device->super.name, task_name,
+                                         lru_gpu_elem, lru_gpu_elem->readers, lru_gpu_elem->super.super.obj_reference_count,
+                                         lru_gpu_elem->original);
                     parsec_atomic_unlock( &oldmaster->lock );
                     goto find_another_data;
                 }
+                /* Check if this copy is the last dangling reference to the oldmaster. This is safe to do as we own one of the data refcounts. */
                 int do_unlock = oldmaster->super.obj_reference_count != 1;
                 parsec_data_copy_detach(oldmaster, lru_gpu_elem, gpu_device->super.device_index);
```

Comment:
This needs to be guaranteed sequenced after the CAS so we need a write memory barrier before the detach.

Reply:
Also, `parsec_data_copy_detach()` doesn't update `data->owner_device`, so this detach can lead to a situation where the `data->device_copies[data->owner_device]` is `NULL`. Is that a behavior we want to support?

Reply:
No, there is no need for write memory barrier. There is a single owner of the copy, the thread that will do the CAS and the detach, and the same that will eventually free the copy. Every other threads is limited to updates of the reader field, if they want to manipulate the copy more, they do it via a message to the owner thread.

I don't see why changing the `owner_device` is needed here, we only change the data_t this copy is associated with, the `owner_device` will remain the same. The detach will clear the `origin` so `data->device_copies[data->owner_device]` will segfault because `data` will be NULL.

Reply:
In `parsec_data_copy_detach` we do `    data->device_copies[device] = copy->older;` (where `data` is `copy->original` before we set `copy->original` to `NULL`. If `copy->older` is NULL, then that `data_t` has a `device_copies[owner_device]` set to NULL?

Reply:
That is a bigger problem overall. 
1. `NULL == copy->older` we have the owner_device pointing to a NULL data. But, we don't know what is legitimately the good version of the remaining copies, so in the generic code we are unable to redirect the `owner_device` to the correct version.
2. `NULL != copy->older` we are setting the `owner_device` to point to an older version of the data that might exists on the device. Not clear why that version would be the correct one.